pkg/kinflate/util.go

```diff
@@ -17,122 +17,232 @@ limitations under the License.
 	"io/ioutil"
+	"os"
 	"path"
+	"path/filepath"
 
 	"github.com/ghodss/yaml"
 
-	"k8s.io/apimachinery/pkg/api/meta"
 	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/strategicpatch"
 	manifest "k8s.io/kubectl/pkg/apis/manifest/v1alpha1"
+	kutil "k8s.io/kubectl/pkg/kinflate/util"
+	"k8s.io/kubectl/pkg/scheme"
 )
 
 const kubeManifestFileName = "Kube-manifest.yaml"
 
-type resource struct {
-	resources  []string
-	configmaps []manifest.ConfigMap
-	secrets    []manifest.Secret
+// loadManifestPkg loads a manifest file and parse it in to the Package object.
+func loadManifestPkg(dirname string) (*manifest.Manifest, error) {
+	bytes, err := ioutil.ReadFile(path.Join(dirname, kubeManifestFileName))
+	if err != nil {
+		return nil, err
+	}
+	var manifest manifest.Manifest
+	// TODO: support json
+	err = yaml.Unmarshal(bytes, &manifest)
+	if err != nil {
+		return nil, err
+	}
+	adjustPathsForManifest(&manifest, []string{dirname})
+	return &manifest, err
 }
 
-type newNameObject struct {
-	newName string
-	obj     runtime.Object
+func populateResourceMap(files []string,
+	m map[kutil.GroupVersionKindName]*unstructured.Unstructured) error {
+	for _, file := range files {
+		_, err := fileToMap(file, m)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
 }
 
-// loadBaseAndOverlayPkg returns:
-// - List of FilenameOptions, each FilenameOptions contains all the files and whether recursive for each base defined in overlay kube-manifest.yaml.
-// - Fileoptions for overlay.
-// - Package object for overlay.
-// - A potential error.
-func loadBaseAndOverlayPkg(f string) ([]*resource, *resource, *manifest.Manifest, error) {
-	overlay, err := loadManifestPkg(path.Join(f, kubeManifestFileName))
+func fileToMap(filename string,
+	into map[kutil.GroupVersionKindName]*unstructured.Unstructured,
+) (map[kutil.GroupVersionKindName]*unstructured.Unstructured, error) {
+	f, err := os.Stat(filename)
 	if err != nil {
-		return nil, nil, nil, err
+		return nil, err
 	}
+	if into == nil {
+		into = map[kutil.GroupVersionKindName]*unstructured.Unstructured{}
+	}
+	switch mode := f.Mode(); {
+	case mode.IsDir():
+		_, err = DirToMap(filename, into)
+	case mode.IsRegular():
+		content, err := ioutil.ReadFile(filename)
+		if err != nil {
+			return nil, err
+		}
 
-	// TODO: support `recursive` when we figure out what its behavior should be.
-	// Recursive: overlay.Recursive
-
-	overlayResource := adjustPathsForConfigMapAndSecret(overlay, []string{f})
-	overlayResource.resources, err = adjustPaths(overlay.Patches, []string{f})
-	if err != nil {
-		return nil, nil, nil, err
+		_, err = kutil.Decode(content, into)
+		if err != nil {
+			return nil, err
+		}
 	}
+	return into, nil
+}
 
-	if len(overlay.Resources) == 0 {
-		return nil, nil, nil, errors.New("expect at least one base, but got 0")
+// DirToMap tries to find Kube-manifest.yaml first in a dir.
+// If not found, traverse all the file in the dir.
```

Comment:
it does more than that - it does recursively finds all instances of Kube-manifest, reads them and merges them all into "into".  

The public entry to this recursion should be a method that does not take "into".  Callers should not be aware of this private accumulator.

Even better, can we tease apart the recursive file traversal from the actual accumulation into a map[kutil.GroupVersionKindName]*unstructured.Unstructured ?
  
Would make it easier to read and test.  The file traversal could return a channel of struct {*Manifest, error} for example.  then something could range over that channel.  the range ends when the channel closes.  or less fancy return an array of manifest, but a channel is actually easier in go.